### PR TITLE
modelling and preprocesiing

### PR DESCRIPTION
**Pull request by RamiroMejia**
Hi guys, I was working on the modelling and there is not much I can do to get better scores. 

I have summarized the modelling/scoring process in four functions that are in the modelling.ipynb file. I think this is the best easiest way to call objects from our pipeline to the documentation.

I  modified the preprocessing of the data. I added/changed some parts of the code in the pre_process_data.py in particular the imputation technique since after research on Kaggle's forums I found that 'X' and 'Y' are coordinates(location). I added a models dictionary to this file that contains the classifiers that we are using to estimate the models. These two objects have been exported to the 'processed' folder as a 'pickle' object.

Now talking about modelling.ipynb the four functions that I created here are:

1.  cv_models(), which takes the model dictionary (loaded using pickle) and X_train and Y_train as arguments. This function returns a pd.Dataframe with the f1scores of all the classifiers. Basically, you just need to load the dictionary that is in the preprocessed folder and use the X_train and y_train come from `data/processed/training_feature.csv`.

2.  best_LR_model, after running 'cv_models()' our best model is LR, so best_LR_model takes X_train and y_train as arguments and does a hyperparameter optimization to search the best model based on the f1 score. I decided that the output of this function is a dictionary that contains the best scores of optimization and the best parameters. If we want to create the best pipeline we can just call the objects from this function using dictionary keys.

Now at this point, there is a part that I didn't automate because I don't know where should it be in the entire workflow. I hard coded the best pipeline `best_pipeline.fit()`. This object is important because is used to create the scores and evaluate the predictions.

3. I created a function call print_scores() that returns the classification report table of the model. It takes 'best_pipeline' as the argument.

4. Finally, the print_confusion_matrix also takes as an argument the best_pipeline.

Sorry for this long message, but I would like to ask you to double check the code.

The next step will be to create a python script that will be easy because everything is a function. Also, you can just call the function to replicate the object. @Thomas let's discuss tomorrow what other ways we can optimize the workflow now that I have a better understanding of 'pickle'.

### CODE REVIEW

File: src/pre_process_data.py [hunk 1]
```diff
@@ -1,71 +1,95 @@
...
-        
-        print(f"Pre-processor is loaded successfully at %s" %(out_path + filename))    
-        
-    except Exception as error:
-        print(f"Error message: %s" %error)
-        print("Error while saving pre-processor!")
-    
-if __name__ == "__main__":
-    main(opt['--out_path'])
+# author: Group 24
```
Comment (thomassiu) [inline src/pre_process_data.py:1]: Hi Ramiro,

Since it is a jupyter notebook I put my comments here instead =)

1. best_LR_model. In case you want to automate it to get the best model automatically, you should then return your results from cv_models as a dict of {"cv_result": results_df, "best_model" model_object}. You can get the highest score from the table and then you can filter out the best model from "models" object.

However that creates another problem, you still have to have specific input for hyperparameterization - technically can be automated with a predefined dict of possible options, HOWEVER I would not recommend to go for that path ... not worth it for now.

So in summary, I think at this moment it is good!
2. I will merge my balance of HOUR back to this script instead.

3. We need to extract more graphs and tables from these steps so that we can answer the following in the report

- What pre-processing actions we have made and why? 
  - A graph that shows the distribution of hours ( HOUR == 0 is crazily high) <=== I will do it
  - A graph that shows distribution after HOUR is balanced <==== I will do it
- A table of the CVs among different models (which we have, just need to be in a png)
- The results (confusion matrix, .... etc)

**Review by thomassiu — APPROVED** [review]
Comments added ...

File: src/pre_process_data.py (hunk 1 above)
Comment (RamiroMejia) [reply src/pre_process_data.py:1]: Answer 1.:

I agree with that, and I was thinking the same but I still don't understand how will use these outputs in the document. We can get outputs from the functions and then used them as arguments for the next functions, but this depends on the report document.

For the hyperparameter optimization I hard coded the parameter grid I did not want to do that but at this point, we don't have many choices 😭, the time complexity is a problem.

Answer 3.

I agree let's work on this together, we have to use the outputs to describe our workflow.

Comment (thomassiu) [reply src/pre_process_data.py:1]: As we talked, hardcode it for now for this project =D, and we can output the pipe_best for people who wants to run the prediction themselves lol

**Comment by thomassiu** [comment]
Please merge it =)

**Review by thomassiu — APPROVED** [review]
